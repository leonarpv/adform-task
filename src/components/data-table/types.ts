import { ReactNode } from 'react'

export type ColumnDefinition<TSource extends {}> = {
    accessor: (source: TSource) => ReactNode
    title: string
}
