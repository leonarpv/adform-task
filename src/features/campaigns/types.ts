export type Campaign = {
    id: number
    name: string
    startDate: string
    endDate: string
    Budget: number
    userId: number
}

export type CampaignUser = {
    id: number
    name: string
}

export type CampaignWithUser = {
    user: CampaignUser | null
} & Campaign

export type CampaignFilters = {
    name?: string
    startDate?: string
    endDate?: string
}
