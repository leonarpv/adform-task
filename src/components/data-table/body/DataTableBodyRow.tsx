import React from 'react'
import { ColumnDefinition } from '../types'
import DataTableBodyCell from './DataTableBodyCell'

type Props<TSource> = {
    columns: ColumnDefinition<TSource>[]
    sourceEntry: TSource
}

const DataTableBodyRow = <TSource extends {}>({
    columns,
    sourceEntry,
}: Props<TSource>) => {
    const cells = columns.map((c, index) => (
        <DataTableBodyCell key={index} column={c} sourceEntry={sourceEntry} />
    ))

    return <tr className="data-table__row">{cells}</tr>
}

export default DataTableBodyRow
