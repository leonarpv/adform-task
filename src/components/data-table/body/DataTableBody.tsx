import React from 'react'
import { ColumnDefinition } from '../types'
import DataTableBodyRow from './DataTableBodyRow'

type Props<TSource> = {
    columns: ColumnDefinition<TSource>[]
    emptyText?: string
    source: TSource[]
}

const DataTableBody = <TSource extends {}>({
    columns,
    emptyText,
    source,
}: Props<TSource>) => {
    const emptyContent = (
        <tr className="data-table__row">
            <td className="data-table__empty-content" colSpan={columns.length}>
                {emptyText || 'No Data'}
            </td>
        </tr>
    )

    const bodyContent =
        source.length === 0
            ? emptyContent
            : source.map((s, index) => (
                  <DataTableBodyRow
                      key={index}
                      columns={columns}
                      sourceEntry={s}
                  />
              ))

    return <tbody className="data-table__body">{bodyContent}</tbody>
}

export default DataTableBody
