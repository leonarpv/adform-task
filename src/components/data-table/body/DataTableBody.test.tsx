import { shallow } from 'enzyme'
import React from 'react'
import { ColumnDefinition } from '../types'
import DataTableBody from './DataTableBody'
import DataTableBodyRow from './DataTableBodyRow'

describe('DataTableBody', () => {
    it('should render properly', () => {
        const columns: ColumnDefinition<string>[] = []
        const source: string[] = []

        const wrapper = shallow(
            <DataTableBody columns={columns} source={source} />
        )

        expect(wrapper).toMatchSnapshot()
    })

    it('should render table body', () => {
        const columns: ColumnDefinition<string>[] = []
        const source: string[] = ['string1', 'string2']

        const wrapper = shallow(
            <DataTableBody columns={columns} source={source} />
        )

        expect(wrapper.find('.data-table__body').exists()).toBe(true)
        expect(wrapper.find(DataTableBodyRow).length).toBe(2)
    })

    it('should render empty table body is source is empty', () => {
        const columns: ColumnDefinition<string>[] = []
        const source: string[] = []

        const wrapper = shallow(
            <DataTableBody columns={columns} source={source} />
        )

        expect(wrapper.find('.data-table__body').exists()).toBe(true)
        expect(wrapper.find(DataTableBodyRow).length).toBe(0)
        expect(wrapper.find('.data-table__empty-content').exists()).toBe(true)
        expect(wrapper.find('.data-table__empty-content').text()).toBe(
            'No Data'
        )
    })

    it('should render provided empty table text', () => {
        const columns: ColumnDefinition<string>[] = []
        const source: string[] = []
        const emptyText = 'emptyText'

        const wrapper = shallow(
            <DataTableBody
                columns={columns}
                source={source}
                emptyText={emptyText}
            />
        )

        expect(wrapper.find('.data-table__body').exists()).toBe(true)
        expect(wrapper.find(DataTableBodyRow).length).toBe(0)
        expect(wrapper.find('.data-table__empty-content').exists()).toBe(true)
        expect(wrapper.find('.data-table__empty-content').text()).toBe(
            emptyText
        )
    })
})
