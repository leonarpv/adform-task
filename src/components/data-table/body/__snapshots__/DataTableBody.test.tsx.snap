// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataTableBody should render properly 1`] = `ShallowWrapper {}`;
